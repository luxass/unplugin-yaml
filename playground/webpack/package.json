{
  "name": "webpack-playground",
  "type": "module",
  "private": true,
  "scripts": {
    "build": "webpack"
  },
  "devDependencies": {
    "unplugin-yaml": "workspace:*",
    "webpack": "^5.38.1",
    "webpack-cli": "^4.7.2"
  }
}
