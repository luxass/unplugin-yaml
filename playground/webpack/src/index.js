import config from "./config.yaml";
import configRaw from "./config.yaml?raw";

console.log(config);
console.log(configRaw);
