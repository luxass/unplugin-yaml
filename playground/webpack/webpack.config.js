// @ts-check
const path = require("node:path");
const YAMLPlugin = require("../../dist/webpack.cjs");

/** @type {import("webpack").Configuration} */
module.exports = {
  entry: "./src/index.js",
  output: {
    filename: "index.js",
    path: path.resolve(__dirname, "dist"),
  },
  plugins: [
    YAMLPlugin(),
  ],
  mode: "production",
  stats: "errors-only",
};
