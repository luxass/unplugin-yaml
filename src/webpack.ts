/**
 * This entry file is for webpack plugin.
 *
 * @module
 */

import { createWebpackPlugin } from "unplugin";
import { unpluginFactory } from "./";

/**
 * Webpack plugin
 *
 * @example
 * ```ts
 * // webpack.config.js
 * module.exports = {
 *  plugins: [require("unplugin-yaml/webpack")()],
 * }
 * ```
 */
export default createWebpackPlugin(unpluginFactory);
