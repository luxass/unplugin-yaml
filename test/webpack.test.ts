import type { Configuration, Stats } from "webpack";
import { readFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import path from "node:path";
import { expect, it } from "vitest";
import { webpack as createWebpack } from "webpack";
import YAMLPlugin from "../src/webpack";
import { removeComments } from "./utils";

const dir = path.join(tmpdir(), "unplugin-yaml-tests");

async function webpack(config: Configuration): Promise<Stats | undefined> {
  return new Promise((resolve, reject) => {
    const compiler = createWebpack({
      optimization: {
        minimize: true,
      },
      output: {
        path: dir,
        filename: `${Date.now()}-bundle.js`,
      },
      mode: "production",
      ...config,
    });

    compiler.run((err, stats) => {
      if (err) {
        reject(err);
      }

      resolve(stats);
    });
  });
}

it("expect yaml import to be a json object", async () => {
  const result = await webpack({
    entry: "./test/fixtures/js-yaml.js",
    plugins: [
      YAMLPlugin(),
    ],
  });

  const json = result?.toJson();
  expect(json).toBeDefined();

  const file = json!.assetsByChunkName!.main;
  const content = await readFile(path.join(dir, file![0]!), "utf-8");

  expect(removeComments(content)).toMatchSnapshot();
});

it("expect yaml import to be a string", async () => {
  const result = await webpack({
    entry: "./test/fixtures/js-yaml-raw.js",
    plugins: [
      YAMLPlugin(),
    ],
  });

  const json = result?.toJson();
  expect(json).toBeDefined();

  const file = json!.assetsByChunkName!.main;
  const content = await readFile(path.join(dir, file![0]!), "utf-8");

  expect(removeComments(content)).toMatchSnapshot();
});

it("expect yml import to be a json object", async () => {
  const result = await webpack({
    entry: "./test/fixtures/js-yml.js",
    plugins: [
      YAMLPlugin(),
    ],
  });

  const json = result?.toJson();
  expect(json).toBeDefined();

  const file = json!.assetsByChunkName!.main;
  const content = await readFile(path.join(dir, file![0]!), "utf-8");

  expect(removeComments(content)).toMatchSnapshot();
});

it("expect yml import to be a string", async () => {
  const result = await webpack({
    entry: "./test/fixtures/js-yml-raw.js",
    plugins: [
      YAMLPlugin(),
    ],
  });

  const json = result?.toJson();
  expect(json).toBeDefined();

  const file = json!.assetsByChunkName!.main;
  const content = await readFile(path.join(dir, file![0]!), "utf-8");

  expect(removeComments(content)).toMatchSnapshot();
});
