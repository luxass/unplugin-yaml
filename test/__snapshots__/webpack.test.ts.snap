// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handle multi document 1`] = `"(()=>{"use strict";console.log([{apiVersion:"batch/v1",kind:"CronJob",metadata:{name:"first-cronjob",namespace:"default"},spec:{schedule:"*/10 * * * *",concurrencyPolicy:"Forbid",successfulJobsHistoryLimit:3,failedJobsHistoryLimit:1,jobTemplate:{spec:{template:{spec:{containers:[{name:"first-job",image:"busybox:latest",command:["/bin/sh","-c",'echo "First job running"; date']}],restartPolicy:"OnFailure"}}}}}},{apiVersion:"batch/v1",kind:"CronJob",metadata:{name:"second-cronjob",namespace:"default"},spec:{schedule:"0 0 * * *",concurrencyPolicy:"Forbid",successfulJobsHistoryLimit:3,failedJobsHistoryLimit:1,jobTemplate:{spec:{template:{spec:{containers:[{name:"second-job",image:"busybox:latest",command:["/bin/sh","-c",'echo "Second job running"; date']}],restartPolicy:"OnFailure"}}}}}}])})();"`;

exports[`handle transforms 1`] = `"(()=>{"use strict";console.log({this:"transformed"})})();"`;

exports[`handle yml > expect yml import to be a json object 1`] = `"(()=>{"use strict";console.log({pluginDir:"./plugins",web:{enabled:!0},logging:{type:"stdout",level:"info"}})})();"`;

exports[`handle yml > expect yml import to be a string 1`] = `"(()=>{"use strict";console.log("pluginDir: ./plugins\\n\\nweb:\\n  enabled: true\\n\\nlogging:\\n  type: stdout\\n  level: info\\n")})();"`;

exports[`handles yaml > expect yaml import to be a json object 1`] = `"(()=>{"use strict";console.log({pluginDir:"./plugins",web:{enabled:!0},logging:{type:"stdout",level:"info"}})})();"`;

exports[`handles yaml > expect yaml import to be a string 1`] = `"(()=>{"use strict";console.log("pluginDir: ./plugins\\n\\nweb:\\n  enabled: true\\n\\nlogging:\\n  type: stdout\\n  level: info\\n")})();"`;
